gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -81,26 +86,42 @@ public BatcherImpl(
       BatchingDescriptor<ElementT, ElementResultT, RequestT, ResponseT> batchingDescriptor,
       UnaryCallable<RequestT, ResponseT> unaryCallable,
       RequestT prototype,
-      BatchingSettings batchingSettings) {
+      BatchingSettings batchingSettings,
+      ScheduledExecutorService executor) {
     this.batchingDescriptor =
         Preconditions.checkNotNull(batchingDescriptor, "batching descriptor cannot be null");
     this.unaryCallable = Preconditions.checkNotNull(unaryCallable, "callable cannot be null");
     this.prototype = Preconditions.checkNotNull(prototype, "request prototype cannot be null");
     this.batchingSettings =
         Preconditions.checkNotNull(batchingSettings, "batching setting cannot be null");
+
+    Preconditions.checkNotNull(executor, "executor cannot be null");
+    final Runnable pushCurrentBatchRunnable =
+        new Runnable() {
+          @Override
+          public void run() {
+            sendBatch();
+          }
+        };
+    long delay = batchingSettings.getDelayThreshold().toMillis();
+    scheduledFuture =
+        executor.scheduleWithFixedDelay(
+            pushCurrentBatchRunnable, delay, delay, TimeUnit.MILLISECONDS);
   }
 
   /** {@inheritDoc} */
   @Override
   public ApiFuture<ElementResultT> add(ElementT element) {
     Preconditions.checkState(!isClosed, "Cannot add elements on a closed batcher");
+    SettableApiFuture<ElementResultT> result = SettableApiFuture.create();
 
-    if (currentOpenBatch == null) {
-      currentOpenBatch = new Batch<>(prototype, batchingDescriptor, batchingSettings);
-    }
+    synchronized (elementLock) {
+      if (currentOpenBatch == null) {
+        currentOpenBatch = new Batch<>(prototype, batchingDescriptor, batchingSettings);
```

Comment:
I think it would be simpler to reason about this if:
- the constructor initialized the currentOpenBatch to an empty batch
- sendBatch was responsible for swapping to a new batch
- add simply add an element to the current batch

Reply:
Agreed this is much simpler. Have updated Batcher with the same approach.